<a href="https://colab.research.google.com/github/Tonry12/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
import os

In [4]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/DPDM2023'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [5]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [6]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

## แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย groupby.apply(list)

In [7]:
data.groupby('InvoiceNo')['StockCode'].apply(list)
#เราดึงมาแค่ รหัสinvoice โดยจะ groupby stockcode มาทำเป็น list

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [8]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [9]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=56cb73f1353f35494babbefbba5f6d5044bf11191373c9bbd1fa51298344914f
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [10]:
from apyori import apriori

In [11]:
apriori(transactions,min_support = 0.065)
#มีอย่างน้อย 6.5% ที่ซื้อของซ้ำๆกันใน transation ก็ถือว่ามี pattern แล้ว

<generator object apriori at 0x7b63aa90b7d0>

In [12]:
list(apriori(transactions,min_support = 0.065))

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [13]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))

In [14]:
len(frequentItemsets)

4

In [15]:
# อ่าน itemset1
frequentItemsets[0]
# มีแค่สินค้าชิ้นเดียว มี support 8.3% ที่ซื้อสินค้านี้
#ซึ่งไม่ได้อะไรมากเท่าไหร่

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [16]:
frequentItemsets[0][0]

frozenset({'22423'})

In [17]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [18]:
frequentItemsets[0][0].__len__()

# itemset นี้มีขนาดเท่ากับ 1

1

In [19]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


มี 2 itemset ที่ผ่าน 0.03

RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,
<t>
<t>

คนที่ซื้อทั้ง 22386, 85099B จาก dataset ทั้งหมดมี 3.21%

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

คนที่ซื้อ 22386 แล้วค่อยซื้อ 85099B มีตั้ง 67.66%


OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),


คนที่ซื้อ 85099B แล้วค่อยซื้อ 22386 มี 39.01%


OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [20]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

### quize 22386, 85099B คือสินค้าอะไร

In [21]:
set(data[data['StockCode']=='22386']['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [22]:
set(data[data['StockCode']=='85099B']['Description'])

{'JUMBO BAG RED RETROSPOT'}

In [23]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

# Homework 1 หา Frequent Item Set และ Association Rules
## ส่งก่อนจันทร์หน้าๆ
จัดกลุ่มตาม invoice
* ของช่วงเวลา Christmas & Newn Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)
* ของช่วงเวลา Christmas & Newn Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* จัดกลุ่มตามประเทศและเดือน


In [32]:
data.groupby('Country').count()

InvoiceNo  StockCode  Description  Quantity  \
Country                                                             
Australia                  1259       1259         1259      1259   
Austria                     401        401          401       401   
Bahrain                      19         19           19        19   
Belgium                    2069       2069         2069      2069   
Brazil                       32         32           32        32   
Canada                      151        151          151       151   
Channel Islands             758        758          758       758   
Cyprus                      622        622          622       622   
Czech Republic               30         30           30        30   
Denmark                     389        389          389       389   
EIRE                       8196       8196         8196      8196   
European Community           61         61           61        61   
Finland                     695        695          695       695   
France                     8557       8557         8557      8557   
Germany                    9495       9495         9495      9495   
Greece                      146        146          146       146   
Hong Kong                   288        288          288       288   
Iceland                     182        182          182       182   
Israel                      297        297          297       297   
Italy                       803        803          803       803   
Japan                       358        358          358       358   
Lebanon                      45         45           45        45   
Lithuania                    35         35           35        35   
Malta                       127        127          127       127   
Netherlands                2371       2371         2371      2371   
Norway                     1086       1086         1086      1086   
Poland                      341        341          341       341   
Portugal                   1519       1519         1519      1519   
RSA                          58         58           58        58   
Saudi Arabia                 10         10           10        10   
Singapore                   229        229          229       229   
Spain                      2533       2533         2533      2533   
Sweden                      462        462          462       462   
Switzerland                2002       2002         2002      2002   
USA                         291        291          291       291   
United Arab Emirates         68         68           68        68   
United Kingdom           495478     495478       494024    495478   
Unspecified                 446        446          446       446   

                      InvoiceDate  UnitPrice  CustomerID  
Country                                                   
Australia                    1259       1259        1259  
Austria                       401        401         401  
Bahrain                        19         19          17  
Belgium                      2069       2069        2069  
Brazil                         32         32          32  
Canada                        151        151         151  
Channel Islands               758        758         758  
Cyprus                        622        622         622  
Czech Republic                 30         30          30  
Denmark                       389        389         389  
EIRE                         8196       8196        7485  
European Community             61         61          61  
Finland                       695        695         695  
France                       8557       8557        8491  
Germany                      9495       9495        9495  
Greece                        146        146         146  
Hong Kong                     288        288           0  
Iceland                       182        182         182  
Israel                        297        297         250  
Italy                 

In [47]:
Asia = data[(data['Country']=='Bahrain') | (data['Country']=='Hong Kong')| (data['Country']=='Singapore') | (data['Country']=='Japan') | (data['Country']=='Israel') | (data['Country']=='Lebanonl') | (data['Country']=='Saudi Arabia') | (data['Country']=='United Arab Emirates')]

In [49]:
Europe = data[(data['Country']!='Bahrain') | (data['Country']!='Hong Kong')| (data['Country']!='Singapore') | (data['Country']!='Japan') | (data['Country']!='Israel') | (data['Country']!='Lebanonl') | (data['Country']!='Saudi Arabia') | (data['Country']!='United Arab Emirates')]

In [48]:
Asia

InvoiceNo StockCode                     Description  Quantity  \
9783      537218     85016   SET OF 6 VINTAGE NOTELETS KIT         6   
9784      537218     21506      FANCY FONT BIRTHDAY CARD,         12   
9785      537218     22716              CARD CIRCUS PARADE        12   
9786      537218     22983             CARD BILLBOARD FONT        12   
9787      537218     22030          SWALLOWS GREETING CARD        12   
...          ...       ...                             ...       ...   
530276   C580832     21731   RED TOADSTOOL LED NIGHT LIGHT        -2   
530277   C580832     23480  MINI LIGHTS WOODLAND MUSHROOMS       -15   
530278   C580832     22466  FAIRY TALE COTTAGE NIGHT LIGHT       -17   
530279   C580832     23084              RABBIT NIGHT LIGHT        -7   
530280   C580832     22621      TRADITIONAL KNITTING NANCY        -1   

             InvoiceDate  UnitPrice  CustomerID Country  
9783    12/05/2010 15:46       2.55     12763.0   Japan  
9784    12/05/2010 15:46       0.42     12763.0   Japan  
9785    12/05/2010 15:46       0.42     12763.0   Japan  
9786    12/05/2010 15:46       0.42     12763.0   Japan  
9787    12/05/2010 15:46       0.42     12763.0   Japan  
...                  ...        ...         ...     ...  
530276  12/06/2011 11:40       1.45     12753.0   Japan  
530277  12/06/2011 11:40       3.29     12753.0   Japan  
530278  12/06/2011 11:40       1.65     12753.0   Japan  
530279  12/06/2011 11:40       1.79     12753.0   Japan  
530280  12/06/2011 11:40       1.45     12753.0   Japan  

[1269 rows x 8 columns]

In [50]:
Europe

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]